lib/happo/snapshot_comparer.rb

```diff
@@ -27,65 +32,89 @@ def compare!
       # work.
       return no_diff if @png_before == @png_after
 
-      array_before = to_array_of_arrays(@png_before)
-      array_after = to_array_of_arrays(@png_after)
-
-      # If the arrays of arrays of colors are identical, we don't need to do any
-      # more work. This might happen if some of the headers are different.
-      return no_diff if array_before == array_after
+      max_height = [@png_after.height, @png_before.height].max
+      max_width = [@png_after.width, @png_before.width].max
 
-      sdiff = Diff::LCS.sdiff(array_before, array_after)
-      number_of_different_rows = 0
+      diff_image = ChunkyPNG::Image.new(max_width, max_height)
 
-      sprite, all_comparisons = initialize_comparison_images(
-        [@png_after.width, @png_before.width].max, sdiff.size)
+      number_of_different_pixels = 0
 
-      sdiff.each_with_index do |row, y|
-        # each row is a Diff::LCS::ContextChange instance
-        all_comparisons.each { |image| image.render_row(y, row) }
-        number_of_different_rows += 1 unless row.unchanged?
+      max_height.times do |y|
+        max_width.times do |x|
+          pixel_before = @png_before.get_pixel(x, y)
+          pixel_after = @png_after.get_pixel(x, y)
+          number_of_different_pixels += 1 unless pixel_before == pixel_after
+          diff_image.set_pixel(x, y, faded_magenta_pixel(
+            pixel_after  || TRANSPARENT,
+            pixel_before || TRANSPARENT
+          ))
+        end
```

Comment:
We would need to write some JS that reads in both images, performs the diff lcs algorithm, and then writes out a new image. I think we can probably use canvas for this, or generate a data URI or something. I don't have all the answers yet but I believe it would be possible.